### PR TITLE
Refactor to use notation-core-go's SignatureEnvelope

### PR DESCRIPTION
**Pull request by priteshbandi**
## Change Log
1. Refactor code to use notation-core-go's SignatureEnvelope
2. Have removed time-stamping functionality to make this PR more readable. Tracking issue: https://github.com/notaryproject/notation-go/issues/78 

## Dependency
1. https://github.com/notaryproject/notation-core-go/pull/15

Because of dependent changes build will fail. Once dependent PR is approved I will update this PR with updated go.sum.

Signed-off-by: Pritesh Bandi <pritesb@amazon.com>

### CODE REVIEW

**Comment by qmuntal** [comment]
@priteshbandi this PR is difficult to review because `signature/jws` files has been moved one level up and Git is not smart enough to detect them as renamed files. Could you leave the directory reorganization for a later PR or split the work in different commits?

**Review by jondonas — APPROVED** [review]

File: plugin/plugin.go [hunk 1]
```diff
@@ -68,22 +68,22 @@ func (DescribeKeyRequest) Command() Command {
...
-	KeySpec notation.KeySpec `json:"keySpec"`
+	KeySpec signer.KeySpec `json:"keySpec"`
 }
 
 // GenerateSignatureRequest contains the parameters passed in a generate-signature request.
 type GenerateSignatureRequest struct {
 	ContractVersion string                 `json:"contractVersion"`
 	KeyID           string                 `json:"keyId"`
-	KeySpec         notation.KeySpec       `json:"keySpec"`
-	Hash            notation.HashAlgorithm `json:"hashAlgorithm"`
```
Comment (priteshbandi) [inline plugin/plugin.go:86]: Should we keep well defined type of hash or an string works?
Decide to use string as we are internally using crypto.Hash to define type.

File: signature/plugin.go [hunk 2]
```diff
@@ -99,81 +99,50 @@ func (s *pluginSigner) generateSignature(ctx context.Context, desc notation.Desc
...
-	// Check keyID is honored.
-	if s.keyID != resp.KeyID {
-		return nil, fmt.Errorf("keyID in generateSignature response %q does not match request %q", resp.KeyID, s.keyID)
+	signReq := signer.SignRequest{
+		Payload:             payloadBytes,
+		PayloadContentType:  signer.PayloadContentTypeV1,
+		SignatureProvider:   psp,
+		SigningTime:         time.Now(),
+		ExtendedSignedAttrs: nil,
+		SigningAgent:        "Notation/1.0.0",
```
Comment (gokarnm) [inline signature/plugin.go:122]: Signing agent is supposed to include external signing plugin's name and version. We can address in another PR, can you open tracking issue.

File: signature/plugin.go [hunk 3]
```diff
@@ -99,81 +99,50 @@ func (s *pluginSigner) generateSignature(ctx context.Context, desc notation.Desc
...
-		return nil, fmt.Errorf("signing algorithm %q in generateSignature response is not supported", resp.SigningAlgorithm)
+	if !opts.Expiry.IsZero() {
+		signReq.Expiry = opts.Expiry
 	}
 
-	// Check certificate chain is not empty.
-	if len(resp.CertificateChain) == 0 {
-		return nil, errors.New("generateSignature response has empty certificate chain")
+	// perform signing plugin signature provider
+	sigEnv, err := signer.NewSignatureEnvelope(signer.MediaTypeJWSJson)
```
Comment (gokarnm) [inline signature/plugin.go:129]: Nit, comment is unclear can you reword it.

File: signature/plugin.go [hunk 4]
```diff
@@ -99,81 +99,50 @@ func (s *pluginSigner) generateSignature(ctx context.Context, desc notation.Desc
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal signing payload: %v", err)
 	}
```
Comment (gokarnm) [inline signature/plugin.go:132]: This seems unrelated to creating a new signature envelope with `NewSignatureEnvelope` , which only throws `UnsupportedSignatureFormatError`.

File: signature/plugin.go [hunk 5]
```diff
@@ -190,7 +159,8 @@ func (s *pluginSigner) mergeConfig(config map[string]string) map[string]string {
 }
 
 func (s *pluginSigner) generateSignatureEnvelope(ctx context.Context, desc notation.Descriptor, opts notation.SignOptions) ([]byte, error) {
-	rawDesc, err := json.Marshal(desc)
+	payload := notation.Payload{TargetArtifact: desc}
+	rawDesc, err := json.Marshal(payload)
 	if err != nil {
 		return nil, err
```
Comment (gokarnm) [inline signature/plugin.go:165]: Should this be similar to line 106?

```suggestion
		return nil, fmt.Errorf("envelope payload can't be marshaled: %w", err)
```

File: signature/plugin.go [hunk 6]
```diff
@@ -190,7 +159,8 @@ func (s *pluginSigner) mergeConfig(config map[string]string) map[string]string {
 }
 
 func (s *pluginSigner) generateSignatureEnvelope(ctx context.Context, desc notation.Descriptor, opts notation.SignOptions) ([]byte, error) {
-	rawDesc, err := json.Marshal(desc)
+	payload := notation.Payload{TargetArtifact: desc}
+	rawDesc, err := json.Marshal(payload)
```
Comment (gokarnm) [inline signature/plugin.go:163]:
```suggestion
	payloadBytes, err := json.Marshal(payload)
```

File: signature/plugin.go [hunk 7]
```diff
@@ -221,50 +190,25 @@ func (s *pluginSigner) generateSignatureEnvelope(ctx context.Context, desc notat
...
+	var signedPayload notation.Payload
+	if err = json.Unmarshal(sigInfo.Payload, &signedPayload); err != nil {
+		return nil, fmt.Errorf("signed envelope payload can't be unmarshaled: %w", err)
 	}
 
-	// Check the the certificate chain conforms to the spec.
-	if err := verifyCertExtKeyUsage(certs[0], x509.ExtKeyUsageCodeSigning); err != nil {
-		return nil, fmt.Errorf("signing certificate does not meet the minimum requirements: %w", err)
+	if !descriptorPartialEqual(desc, signedPayload.TargetArtifact) {
+		return nil, errors.New("descriptor subject has changed")
```
Comment (gokarnm) [inline signature/plugin.go:209]: Envelope generator should not add any additional top level payload attributes, we should verify that.

File: signature/signer.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+	"errors"
+	"fmt"
+
+	"github.com/notaryproject/notation-core-go/signer"
+	"github.com/notaryproject/notation-go"
+	"github.com/notaryproject/notation-go/plugin"
+)
+
+// NewSignerFromFiles creates a signer from key, certificate files
+// TODO: Add tests for this method.
```
Comment (gokarnm) [inline signature/signer.go:17]: I'd create a tracking issue and add in the comment.

File: signature/signer.go [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+			CertificateChain: rawCerts,
+		}, nil
+	}
+	return nil, plugin.RequestError{
+		Code: plugin.ErrorCodeGeneric,
+		Err:  fmt.Errorf("command %q is not supported", req.Command()),
+	}
+}
+
+// called from plugin.go
```
Comment (gokarnm) [inline signature/signer.go:134]: Remove dead code?

File: signature/verifier.go [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+	"github.com/notaryproject/notation-go"
+)
+
+// // maxTimestampAccuracy specifies the max acceptable accuracy for timestamp.
+// const maxTimestampAccuracy = time.Minute
+
+// Verifier verifies artifacts against JWS signatures.
+type Verifier struct {
+	TrustedCerts []*x509.Certificate
+}
```
Comment (gokarnm) [inline signature/verifier.go:20]: Do we still need this type?

**Review by gokarnm — CHANGES_REQUESTED** [review]

File: signature/verifier.go (hunk 10 above)
Comment (priteshbandi) [reply signature/verifier.go:20]: Yes to keep existing verification code working we need this. This should be removed when we have implemented trust store.

File: signature/signer.go (hunk 9 above)
Comment (priteshbandi) [reply signature/signer.go:134]: It’s the code which contains old time-stamping logic. Technically, it is a dead code but keeping it because it can be used as reference when implementing timestamping in notation-core-go

File: signature/plugin.go (hunk 2 above)
Comment (priteshbandi) [reply signature/plugin.go:122]: Added todo in code...Will open one uber issue with all TODOs.

File: signature/plugin.go (hunk 7 above)
Comment (priteshbandi) [reply signature/plugin.go:210]: Added TODO.

**Review by gokarnm — APPROVED** [review]
LGTM! Tracking open issues in https://github.com/notaryproject/notation-go/issues/80 .

**Comment by priteshbandi** [comment]
Reopening

**Review by jondonas — APPROVED** [review]
